[pilot] Do not use spaceship if both skip_waiting_for_build_processing and apple_id are set

janpio opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
With the new 2FA rules by Apple, more and more users are looking for a way to use app specific passwords to upload to the App Store. 

### Description
As iTunes Transporter does accept app specific passwords, and is used by `pilot` to upload the binary, we can achieve this with a few tiny changes in the program logic (do not log in via Spaceship if there is no reason to. This is the case if both `skip_waiting_for_build_processing` and `apple_id` are set).

closes #14258
closes #14723

janpio committed:
make pilot not use spaceship if both skip_waiting_for_build_processing and apple_id are set

janpio committed:
make rubocop happy

janpio committed:
use options instead of config (assumption: they are actually identical)

janpio committed:
still have to fill @config somehow when start is not called (UGLY!)

janpio committed:
prepare test a bit

janpio committed:
2 tests that make sure that the spaceship login method is only called when the 2 options are _not_ set

janpio committed:
make rubocop happy

janpio committed:
add missing options param

janpio committed:
add missing options param

janpio committed:
add some docs

janpio commented on pilot/lib/pilot/build_manager.rb:
```diff
@@ -10,11 +10,12 @@
     def upload(options)
-      start(options)
+      start(options) unless options[:skip_waiting_for_build_processing] && options[:apple_id]
+      @config = options if options[:skip_waiting_for_build_processing] && options[:apple_id] # ugly as hell, we want to do that differently
```
This is ugly as hell, but the only way to have `@config` be filled as it would be in `start`. I want to avoid calling `start()`, as it triggers the spaceship login.

janpio replied:
Actually, thinking about it, doing this here manually might _break_ further calls to `start` because it will prevent the spaceship login to happen later if the presence of `@config` is used as an indicator if a login has happened before or not.

Not a problem right now as we don't have any further calls of `start` or anything that then needs the spaceship login in this case, it is still ugly.

What would a better way to acheive the same thing be? Add a `spaceship_login_happened` bool to use instead of abusing `@config`?

Commit:
A few small improvements

joshdholtz reviewed (commented):
@janpio This is a cool PR 🙌 I rebased and pushed some changes on top here to clean up that start/login logic that you commented on that needed improvement. Let me know your thoughts!

janpio commented on pilot/lib/pilot/build_manager.rb:
```diff
@@ -50,15 +51,17 @@ def upload(options)
         return
       end
 
+      login_if_not_logged_in
```
That's smart.

janpio commented on pilot/lib/pilot/build_manager.rb:
```diff
@@ -218,7 +225,8 @@ def should_update_localized_build_information?(options)
     def transporter_for_selected_team(options)
       generic_transporter = FastlaneCore::ItunesTransporter.new(options[:username], nil, false, options[:itc_provider])
-      return generic_transporter unless options[:itc_provider].nil? && Spaceship::Tunes.client.teams.count > 1
+      return generic_transporter if options[:itc_provider] || Spaceship::Tunes.client.nil?
```
What's this change about?

joshdholtz replied:
@janpio `Spaceship::Tunes.client.teams.count > 1` would 💣 if not logged in so checking that on the following line separately 😊

janpio commented:
Ok, your changes look good to me @joshdholtz. From my side this is good to merge.

snatchev commented on pilot/lib/pilot/build_manager.rb:
```diff
@@ -49,6 +51,8 @@ def upload(options)
         return
       end
 
+      login_if_not_logged_in
```
Code style nitpick: methods with no input and no outputs give me pause. Why does this exist if it's only called once (and it doesn't appear to be used in mocking or anything like that) ?

janpio replied:
@joshdholtz, your code.

janpio replied:
ping @joshdholtz

snatchev commented on pilot/spec/build_manager_spec.rb:
```diff
@@ -221,4 +223,61 @@
+        fake_build_manager.upload(upload_options)
+      end
+
+      it "is executed when skip_waiting_for_build_processing and apple_id are not set" do
```
There are no expectations in this spec. Perhaps change some of the `allow`s to `expect` ?

janpio replied:
Silly question: What's the difference?

snatchev replied:
`expect` is similar to an `allow`, but it will fail if that method does _not_ get called by the end of the spec. Just by having `allow`, the spec will either throw an error or pass, but it cannot fail.

janpio replied:
Ah ok, then this really makes sense to change here. 
I built this test up from nothing until it let the code execute enough to pass.

janpio replied:
I added an appropriate `expect` to each of the two specs: Once that the `Manager.login` method (which triggers a spaceship login) is _NOT_ called, and once that it is called at least once.

snatchev commented on pilot/spec/build_manager_spec.rb:
```diff
@@ -221,4 +223,61 @@
+      end
+
+      it "is not executed when skip_waiting_for_build_processing and apple_id are set" do
+        fake_build_manager.upload(upload_options)
```
I am unclear on how we assert that `upload` is not executed here.

janpio replied:
This test is: When using the upload, no login is executed, under special circumstances.
Although this is only verified by all the other stuff from line 254 and following not being present.

janpio replied:
Taken care of with the `expect.not to.`

snatchev commented:
Heyo, who's the point person on this? @janpio or @joshdholtz? Couldnt quite tell from the history.

janpio commented:
I had the idea and implemented a first version, with some pretty complicated code. Then @joshdholtz fortunately jumped in and refactored it to something a lot more readable via https://github.com/fastlane/fastlane/pull/14456/commits/68eddca23178e22c768152de96487e726e6005cf

I have on my TODO to rewrite the tests to take care of the things your commented on, and I @ mentioned @joshdholtz in your comment that applied to something that he wrote.

janpio committed:
rename and clarify specs

janpio committed:
add explicit expect and !expect
clean and resort allows

janpio committed:
rubocop

janpio committed:
correct order

janpio commented:
Rebased so CI can actually run (Circle removed our old Xcode 9.0.0 we were using).

Also reworked the tests to address valid and great feedback from @snatchev. Now uses `expect` to make sure the login method that triggers the spaceship login is called or NOT called correctly (which this PR is all about).

I also created https://github.com/fastlane/fastlane/pull/14723 with _just_ the tests of this, to show that the current codebase always calls the spaceship login and thus fails the test (in the expected way).

Commit:
Removed the login_if_needed method and replaced with better code and comments

joshdholtz reviewed (approved):
This is 💯 to me!! LGTM
